torch_geometric/nn/pool/asap.py

```diff
@@ -0,0 +1,119 @@
+    Graph Representations" <https://arxiv.org/abs/1911.07979>`_ paper.
+
+    Args:
+        in_channels (int): Size of each input sample.
+        ratio (float, optional): Graph pooling ratio, which is used to compute
+            :math:`k = \lceil \mathrm{ratio} \cdot N \rceil`.
+            (default: :obj:`0.5`)
+        GNN (torch.nn.Module, optional): A graph neural network layer for
+            using intra-cluster properties.
+            Especially helpful for graphs with higher degree of neighborhood
+            (one of :class:`torch_geometric.nn.conv.GraphConv`,
+            :class:`torch_geometric.nn.conv.GCNConv` or
+            any GNN which supports the :obj:`edge_weight` parameter).
+            (default: :obj:`None`)
+        dropout (float, optional): Dropout probability of the normalized
+            attention coefficients which exposes each node to a stochastically
+            sampled neighborhood during training. (default: :obj:`0`)
+        negative_slope (float, optional): LeakyReLU angle of the negative
+            slope. (default: :obj:`0.2`)
+        add_self_loops (bool, optional): If set to :obj:`True`, will add self
+            loops to the new graph connectivity. (default: :obj:`True`)
+        **kwargs (optional): Additional parameters for initializing the
+            graph neural network layer.
+    """
+    def __init__(self, in_channels, ratio=0.5, GNN=None, dropout=0,
+                 negative_slope=0.2, add_self_loops=True, **kwargs):
+        super(ASAPooling, self).__init__()
+
+        self.in_channels = in_channels
+        self.ratio = ratio
+        self.negative_slope = negative_slope
+        self.dropout = dropout
+        self.GNN = GNN
+        self.add_self_loops = add_self_loops
+
+        self.lin = Linear(in_channels, in_channels)
+        self.att = Linear(2 * in_channels, 1)
+        self.gnn_score = LEConv(self.in_channels, 1)
+        if self.GNN is not None:
+            self.gnn_intra_cluster = GNN(self.in_channels, self.in_channels,
+                                         **kwargs)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        self.lin.reset_parameters()
+        self.att.reset_parameters()
+        self.gnn_score.reset_parameters()
+        if self.GNN is not None:
+            self.gnn_intra_cluster.reset_parameters()
+
+    def forward(self, x, edge_index, edge_weight=None, batch=None):
+        N = x.size(0)
+
+        if batch is None:
+            batch = edge_index.new_zeros(x.size(0))
+
+        x = x.unsqueeze(-1) if x.dim() == 1 else x
+
+        x_pool = x
+        if self.GNN is not None:
+            x_pool = self.gnn_intra_cluster(x=x, edge_index=edge_index,
+                                            edge_weight=edge_weight)
+
+        x_pool_j = x_pool[edge_index[0]]
+        x_q = scatter(x_pool_j, edge_index[1], dim=0, reduce='max')
+        x_q = self.lin(x_q)[edge_index[1]]
```

Comment:
It does for undirected graphs, and it is in line with default message flow "source_to_target" where edge_index[0] refers to x_j.

Reply:
`x_pool_j` denotes the neighborhood of a source node which will form a cluster. So shouldn't `edge_index[1]` be used for finding `x_pool_j` ?

Reply:
`edge_index[0]` denote the source nodes, and `edge_index[1]` denote the target nodes. Since we want to aggregate into the target nodes, `x_j` refers to `x[edge_index[0]]`.

Reply:
In ASAP, the cluster formation starts by treating each node as a centre of a cluster and using its neighborhood nodes as its cluster member. So isn't it like a GCN, where we want to aggregate into the `source nodes`?

Reply:
I think the message flow should be similar to the one used when applying `self.GNN`, right? Of course, we can change it back if you think it is the right way to do so.

Reply:
However, all GNNs in PyG aggregate into the target nodes by default.

Reply:
I get it now! Makes sense to me :)